pkg/receiver/telegrafreceiver/receiver.go

```diff
@@ -105,6 +110,37 @@ func (r *telegrafreceiver) Start(ctx context.Context, host component.Host) error
 	return err
 }
 
+// Consume metrics and retry on recoverable errors
+func (r *telegrafreceiver) consumeWithRetry(ctx context.Context, metrics pdata.Metrics) error {
+	constantBackoff := backoff.WithMaxRetries(backoff.NewConstantBackOff(r.consumeRetryDelay), r.consumeMaxRetries)
+
+	// retry handling according to https://github.com/open-telemetry/opentelemetry-collector/blob/master/component/receiver.go#L45
+	err := backoff.RetryNotify(
+		func() error {
+			// we need to check for context cancellation here
+			select {
+			case <-ctx.Done():
+				return backoff.Permanent(errors.New("closing"))
+			default:
+			}
```

Comment:
Be wary of the fact that this might actually go through the `default` case even though `<-ctx.Done()` can also proceed.

Reply:
Yeah, but that's ok in this case. I'm basically making sure that this exits cleanly during shutdown. At worst it'll get a permanent error from the rest of the pipeline and exit that way, or spit out a stack trace.